minkasi/needlet/needlet.py

```diff
@@ -37,6 +37,216 @@
 else:
     from typing_extensions import Self
 
+################################################################
+#                       Basis Functions                        #
+################################################################
+
+
+def __Standard(
+    k_arr: NDArray[np.floating], j: int, B: np.floating
+) -> NDArray[np.floating]:
+    """
+    Driver function for generating standard needlets
+
+    Parameters
+    ----------
+    k_arr : NDArray[np.floating]
+        Array of k at which to evaluate needlet.
+    j : int
+        The index of the needlet.
+    B : np.floating
+        Parameter which sets the needlet width
+
+    Returns
+    -------
+    to_ret : NDArray[np.floating]
+        1D standard needlet response function for index j
+    """
+
+    def __f_need(t):
+        """Auxiliar function f to define the standard needlet"""
+        if t <= -1.0:
+            return 0.0
+        elif t >= 1.0:
+            return 0.0
+        else:
+            return np.exp(1.0 / (t**2.0 - 1.0))
+
+    def __psi(u):
+        """Auxiliar function psi to define the standard needlet"""
+        return integrate.quad(__f_need, -1, u)[0] / integrate.quad(__f_need, -1, 1)[0]
+
+    def __phi(q, B):
+        """Auxiliar function phi to define the standard needlet"""
+        B = float(B)
+        if q < 0.0:
+            raise ValueError("The multipole should be a non-negative value")
+        elif q <= 1.0 / B:
+            return 1.0
+        elif q >= 1.0:
+            return 0
+        else:
+            return __psi(1.0 - (2.0 * B / (B - 1.0) * (q - 1.0 / B)))
+
+    xi = k_arr / B**j
+    b2 = __phi(xi / B, B) - __phi(xi, B)
+    return np.max([0.0, b2])
+
+
+def Standard(
+    k_arr: NDArray[np.floating], js: NDArray[int], B: np.floating
+) -> NDArray[np.floating]:
+    """
+    Helper function for constructing standard needlet basis.
+    TODO: this might be cleaner with np.vectorize.
+
+    Parameters
+    ----------
+    k_arr : NDArray[np.floating]
+        Array of k at which to evaluate needlets.
+    js : NDArray[int]
+        The indicies of the needlets.
+    B : np.floating
+        Parameter which sets the needlet width
+
+    Returns
+    -------
+    to_ret : NDArray[np.floating]
+        1D standard needlet response functions
+    """
+
+    to_ret = []
```

Comment:
I think we know how long this is ahead of time? Its will have shape `(len(js), len(k_arr))`, you should just prealloc with `np.empty` to speed things up.